cmd/drand-gossip-relay/lp2p/ctor.go

```diff
@@ -0,0 +1,117 @@
+package lp2p
+
+import (
+	"context"
+	"crypto/rand"
+	"fmt"
+	mrand "math/rand"
+	"time"
+
+	"github.com/ipfs/go-datastore"
+	"github.com/ipfs/go-datastore/namespace"
+	logging "github.com/ipfs/go-log/v2"
+	"github.com/libp2p/go-libp2p"
+	"github.com/libp2p/go-libp2p-core/crypto"
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	dht "github.com/libp2p/go-libp2p-kad-dht"
+	pubsub "github.com/libp2p/go-libp2p-pubsub"
+	record "github.com/libp2p/go-libp2p-record"
+	routing "github.com/libp2p/go-libp2p-routing"
+	libp2ptls "github.com/libp2p/go-libp2p-tls"
+	ma "github.com/multiformats/go-multiaddr"
+	xerrors "golang.org/x/xerrors"
+)
+
+var (
+	log       = logging.Logger("lp2p")
+	privDsKey = datastore.NewKey("p2pPrivKey")
+)
+
+func PubSubTopic(nn string) string {
+	return fmt.Sprintf("/drand/pubsub/v0.0.0/%s", nn)
+}
+
+func ConstructHost(ds datastore.Datastore, priv crypto.PrivKey, listenAddr string, bootstrap []ma.Multiaddr) (host.Host, *dht.IpfsDHT, *pubsub.PubSub, error) {
+	var idht *dht.IpfsDHT
+	dhtDs := namespace.Wrap(ds, datastore.NewKey("/dht"))
+
+	addrInfos, err := peer.AddrInfosFromP2pAddrs(bootstrap...)
+	if err != nil {
+		fmt.Printf("%+v", bootstrap)
+		return nil, nil, nil, xerrors.Errorf("parsing addrInfos: %+v", err)
+	}
+
+	ctx := context.TODO()
+	h, err := libp2p.New(ctx,
+		libp2p.ListenAddrStrings(listenAddr),
+		libp2p.Identity(priv),
+		libp2p.Security(libp2ptls.ID, libp2ptls.New),
+		libp2p.Routing(func(h host.Host) (routing.PeerRouting, error) {
+			var err error
+			idht, err = dht.New(ctx, h,
+				dht.Mode(dht.ModeServer),
```

Comment:
probably auto?

Reply:
True, I focused too much on the relay itself.